# AC 221: Critical Thinking in Data Science

## Problem Set 1, Part 2
**February 27, 2020**<br/>
**Harvard University**<br/>
**Spring 2020**<br/>

<hr style="height:2pt">

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
quasi = pd.read_csv('data/final_quasi-identifiers.csv')
quasi_list = list(quasi['variable'].values)

# Drop user_id
quasi_list.pop(1)

'user_id'

In [3]:
df = pd.read_csv('APCOMP221Pset1.csv').drop(['user_id', 'username', 'ip', 
                                             'un_major_region', 'un_economic_group', 
                                             'un_developing_nation', 'un_special_region',
                                            'latitude', 'longitude', 'profile_country'], axis=1)

na_replacements = {
    'course_id': 'NA',
    'cc_by_ip': 'NA',
    'countryLabel': 'NA',
    'continent': 'NA',
    'city': 'NA',
    'region': 'NA',
    'subdivision': 'NA',
    'postalCode': 'NA',
    'LoE': 'NA',
    'YoB': -1,
    'gender': 'NA',
    'nforum_posts': 0,
    'nforum_votes': 0,
    'nforum_endorsed': 0,
    'nforum_threads': 0,
    'nforum_comments': 0,
    'nforum_pinned': 0,
    'nforum_events': 0,
    'mode': 'NA',
    'email_domain': 'NA'
}
df = df.fillna(na_replacements)
df

/home/justin/miniconda3/envs/ac221/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3062: DtypeWarning: Columns (49,50,64) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


course_id  registered  viewed explored  certified  \
0        HarvardX/PH525.1x/1T2018        True   False      NaN      False   
1        HarvardX/PH525.1x/1T2018        True    True     True      False   
2        HarvardX/PH525.1x/1T2018        True    True    False      False   
3        HarvardX/PH525.1x/1T2018        True    True    False      False   
4        HarvardX/PH525.1x/1T2018        True    True     True      False   
...                           ...         ...     ...      ...        ...   
1141730  HarvardX/PH525.5x/2T2018        True   False      NaN      False   
1141731  HarvardX/PH525.5x/2T2018        True   False      NaN      False   
1141732  HarvardX/PH525.5x/2T2018        True   False      NaN      False   
1141733  HarvardX/PH525.5x/2T2018        True   False      NaN      False   
1141734  HarvardX/PH525.5x/2T2018        True   False      NaN      False   

         completed cc_by_ip    countryLabel      continent  \
0            False       GB  United Kingdom         Europe   
1            False       US   United States       Americas   
2            False       NA              NA             NA   
3            False       US   United States  North America   
4            False       HU         Hungary         Europe   
...            ...      ...             ...            ...   
1141730      False       NA              NA             NA   
1141731      False       IN           India           Asia   
1141732      False       CN           China           Asia   
1141733      False       BR          Brazil       Americas   
1141734      False       IN           India           Asia   

                        city  ... roles_isInstructor roles_isStaff  \
0              Middlesbrough  ...                NaN           NaN   
1                    Seattle  ...                NaN           NaN   
2                         NA  ...                NaN           NaN   
3                         NA  ...                NaN           NaN   
4                   Budapest  ...                NaN           NaN   
...                      ...  ...                ...           ...   
1141730                   NA  ...                NaN           NaN   
1141731                Delhi  ...                NaN           NaN   
1141732              Beijing  ...                NaN           NaN   
1141733  Almirante Tamandaré  ...                NaN           NaN   
1141734                Anand  ...                NaN           NaN   

        roles_isCCX roles_isFinance  roles_isLibrary roles_isSales  \
0               NaN             NaN              NaN           NaN   
1               NaN             NaN              NaN           NaN   
2               NaN             NaN              NaN           NaN   
3               NaN             NaN              NaN           NaN   
4               NaN             NaN              NaN           NaN   
...             ...             ...              ...           ...   
1141730         NaN             NaN              NaN           NaN   
1141731         NaN             NaN              NaN           NaN   
1141732         NaN             NaN              NaN           NaN   
1141733         NaN             NaN              NaN           NaN   
1141734         NaN             NaN              NaN           NaN   

         forumRoles_isAdmin  forumRoles_isCommunityTA forumRoles_isModerator  \
0                       NaN                       NaN                    NaN   
1                       NaN                       NaN                    NaN   
2                       NaN                       NaN                    NaN   
3                       NaN                       NaN                    NaN   
4                       NaN                       NaN                    NaN   
...                     ...                       ...                    ...   
1141730                 NaN                       NaN                    NaN   
1141731                 NaN     

In [4]:
def percent_remaining(k):
    groups = df.groupby(quasi_list).size()
    groups_below_threshold = groups[groups < k]
    return (1 - groups_below_threshold.sum() / df.shape[0]) * 100

percent_remaining(3)
#groups_below_threshold.sum()

13.448304554033996

In [5]:
percent_remaining(4)

11.527018090887985

In [6]:
percent_remaining(5)

10.348110551047307

### Part 2

In [82]:
completion_counts = df['completed'].value_counts(dropna=False)
print(f'Total completion rate: {completion_counts[True] / df.shape[0] * 100:.5}')

Total completion rate: 2.4648


In [95]:
groups = df.groupby(quasi_list).size()
groups_meeting_threshold = groups[groups >= k]

#return (1 - groups_below_threshold.sum() / df.shape[0]) * 100

False    1141735
Name: course_id, dtype: int64

## TODO

In [97]:
groups_meeting_threshold.index[0]

('HarvardX/1368.1x/2T2016',
 'NA',
 'NA',
 'NA',
 'NA',
 'NA',
 'NA',
 'NA',
 'NA',
 -1.0,
 'NA',
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 'honor',
 'gmail.com')

## Question 4

In [104]:
def n_synthetic_required_for(k):
    groups = df.groupby(quasi_list).size()
    groups_below_threshold = groups[groups < k]
    return (k - groups_below_threshold).sum()

In [105]:
n_synthetic_required_for(5)

3837543

In [108]:
import random

In [139]:
col_counts = df['cc_by_ip'].value_counts().to_dict()
col_counts

{'US': 308402,
 'NA': 184426,
 'IN': 77301,
 'GB': 40781,
 'BR': 38065,
 'CA': 36779,
 'CN': 25987,
 'AU': 20788,
 'DE': 18388,
 'MX': 18350,
 'ES': 16420,
 'FR': 14932,
 'CO': 12049,
 'RU': 11443,
 'NL': 10775,
 'TR': 10220,
 'PK': 9650,
 'EG': 9464,
 'GR': 9043,
 'PH': 8837,
 'JP': 8688,
 'SG': 8635,
 'NG': 8515,
 'IT': 8221,
 'KR': 8001,
 'ID': 7901,
 'HK': 7753,
 'PL': 7569,
 'UA': 7143,
 'TW': 6798,
 'PE': 6416,
 'VN': 6030,
 'AR': 5990,
 'ZA': 5342,
 'CL': 5299,
 'MY': 5288,
 'BE': 5280,
 'PT': 5273,
 'RO': 5140,
 'CH': 4539,
 'SA': 4485,
 'SE': 4414,
 'TH': 4356,
 'AE': 3952,
 'IE': 3909,
 'MA': 3589,
 'NZ': 3510,
 'IL': 3258,
 'BD': 3041,
 'KE': 2900,
 'DK': 2820,
 'NO': 2771,
 'AT': 2765,
 'HU': 2726,
 'EC': 2712,
 'GH': 2683,
 'CZ': 2534,
 'RS': 2506,
 'IR': 2275,
 'FI': 2084,
 'VE': 1946,
 'BG': 1919,
 'NP': 1817,
 'DZ': 1729,
 'CR': 1694,
 'HR': 1532,
 'LK': 1531,
 'LT': 1355,
 'DO': 1344,
 'TN': 1302,
 'PR': 1190,
 'UG': 1178,
 'ET': 1156,
 'GT': 1155,
 'JO': 1151,
 'AL': 

In [119]:
pd.Series(random.choices(col_counts.index.values, weights=col_counts.values, k=int(1.1e6))).value_counts()

US    297305
NA    178095
IN     74130
GB     39360
BR     36850
       ...  
PW         3
AX         2
FK         1
ST         1
SM         1
Length: 224, dtype: int64

## Question 5

We've decided to do generalization on each of the quasi-identifiers first, and then suppress any additional records that need it. This isn't a particularly sophisticated technique, and it's certainly not the optimization that was performed in Angiuli and Waldo's "Statistical Tradeoffs ...".

Here are the generalization strategies we're using:
  * Drop `cc_by_ip` as it's largely redundant with `countryLabel`
  * Starting with the most specific, set any geographic field with fewer than 5000 records to "NA"
  * Change `YoB` to `birth_decade` which is the first three digits of the year
  * Change all the `nforum` varibles to booleans indicating whether they are greater than zero
  * Change `email_domain` to only the first three letters
  * Change `LoE` to `attended_college`; a boolean indicating if the student went to college
  
Changing the names of some of these attributes means we need a new set of quasi-identifiers to work from:

In [82]:
GENERALIZED_QUASI_IDS = [
    'postalCode',
    'city',
    'subdivision',
    'region',
    'countryLabel',
    'continent',
    'email_domain',
    'course_id',
    'mode',
    'gender',
    'nforum_posts_>0',
    'nforum_votes_>0',
    'nforum_endorsed_>0',
    'nforum_threads_>0',
    'nforum_comments_>0',
    'nforum_pinned_>0',
    'nforum_events_>0',
    'birth_decade',
    'attended_college'
]

### Useful Functions

Here we write all our functions for generalization, suppression, and presentation of results.

In [83]:
def generalize_geo(df, min_per_geo=5000):
    g = df.copy()
    GEO_HIERARCHY = [
        'postalCode',
        'city',
        'subdivision',
        'region',
        'countryLabel',
        'continent']
    g = g.drop('cc_by_ip', axis=1)
    for field in GEO_HIERARCHY:
        value_counts = g[field].value_counts()
        values_to_drop = set(value_counts[value_counts < min_per_geo].index.values)
        g.loc[g[field].isin(values_to_drop), field] = na_replacements[field]
    return g

In [84]:
def generalize_education(df):
    g = df.copy()
    # Associate is sometimes not included as college, but we are including it.
    college_levels = set(['p', 'p_se', 'p_oth', 'm', 'b', 'a'])
    g['attended_college'] = g['LoE'].isin(college_levels)
    return g.drop('LoE', axis=1)

In [85]:
def generalize_birth_year(df):
    g = df.copy()
    g['birth_decade'] = df['YoB'].apply(lambda y: str(y)[:3] + '*' if y>-1 else 'NA')
    return g.drop('YoB', axis=1)

In [86]:
def generalize_forum_actions(df):
    g = df.copy()
    fields = ['nforum_posts',
        'nforum_votes',
        'nforum_endorsed',
        'nforum_threads',
        'nforum_comments',
        'nforum_pinned',
        'nforum_events']
    for field in fields:
        g[field + '_>0'] = g[field] > 0
        g = g.drop(field, axis=1)
    return g

In [87]:
def generalize_email_domains(df):
    g = df.copy()
    g['email_domain'] = g['email_domain'].apply(lambda e: e[:3] + '*')
    return g

In [88]:
def generalize_df(df):
    return generalize_geo(
        generalize_birth_year(
        generalize_education(
        generalize_email_domains(
        generalize_forum_actions(df)))))

In [111]:
def suppress_to_k_anonymize(df, k):
    return df.groupby(GENERALIZED_QUASI_IDS).filter(lambda g: g.shape[0] >= k)

In [73]:
def status_count(df, status):
    return df.query(status + ' == True').shape[0]

def status_proportion(df, status):
    return status_count(df, status) / df.shape[0] * 100

def print_status_stats(df):
    print(f'Total students: {df.shape[0]}\n'
          f'Students explored: {status_count(df, "explored")}'
          f' - {status_proportion(df, "explored"):.2f}%\n'
          f'Students completed: {status_count(df, "completed")}'
          f' - {status_proportion(df, "completed"):.2f}%')

### Raw Dataset Stats

In [75]:
print_status_stats(df)

Total students: 1141735
Students explored: 122258 - 10.71%
Students completed: 28142 - 2.46%


### 3-Anonymous Results

In [ ]:
generalized = generalize_df(df)

In [112]:
k3_df = suppress_to_k_anonymize(generalized, 3)

In [114]:
print_status_stats(k3_df)

Total students: 596580
Students explored: 41138 - 6.90%
Students completed: 4120 - 0.69%


### 4-Anonymous Results

In [115]:
k4_df = suppress_to_k_anonymize(generalized, 4)

In [116]:
print_status_stats(k4_df)

Total students: 531984
Students explored: 34513 - 6.49%
Students completed: 3102 - 0.58%


### 5-Anonymous Results

In [118]:
k5_df = suppress_to_k_anonymize(generalized, 5)

In [119]:
print_status_stats(k5_df)

Total students: 487536
Students explored: 30230 - 6.20%
Students completed: 2492 - 0.51%


### Observations

We can see that there's a steep dropoff in the number and proportion of students that completed and explored the class between the raw dataset and the 3-anonymous set. Specifically, in the raw data 2.46% of students completed, whereas in the 3-anonymous dataset, only 0.69% of students completed. As we increase the level of k-anonymity, there are fewer students overall and the proportion of students that completed and explored continues to drop off. This demonstrates that making a dataset k-anonymous may seriously impact the statistical validity of the dataset.

## Question 6

After reviewing all the field, we believe the following could be sensitive: `completed`, `grade`, `nproblem_check`, `nshow_answer`, `cert_status`. To check for the l-diversity, we're taking the minimum number of unique values per field for each quasi-identifier group.

In [120]:
def calc_l_diversity(df, field):
    return df.groupby(GENERALIZED_QUASI_IDS).apply(lambda g: g[field].nunique(dropna=False)).min()

In [122]:
sensitive_fields = ['completed', 'grade', 'nproblem_check', 'nshow_answer', 'cert_status']
k3_l_diversity = {field: calc_l_diversity(k3_df, field) for field in sensitive_fields}
k4_l_diversity = {field: calc_l_diversity(k4_df, field) for field in sensitive_fields}
k5_l_diversity = {field: calc_l_diversity(k5_df, field) for field in sensitive_fields}

### 3-Anonymous l-diversity Results

In [123]:
k3_l_diversity

{'completed': 1,
 'grade': 1,
 'nproblem_check': 1,
 'nshow_answer': 1,
 'cert_status': 1}

### 4-Anonymous l-diversity Results

In [124]:
k4_l_diversity

{'completed': 1,
 'grade': 1,
 'nproblem_check': 1,
 'nshow_answer': 1,
 'cert_status': 1}

### 5-Anonymous l-diversity Results

In [125]:
k5_l_diversity

{'completed': 1,
 'grade': 1,
 'nproblem_check': 1,
 'nshow_answer': 1,
 'cert_status': 1}

### Observations

From the above results, we can see that after all our work, none of our datasets rise above 1-diversity for any of the fields we consider sensitive. The improve this, we would need to either consider suppressing entire quasi-identifier columns or creating synthetic records to add diversity to each quasi-identifier group.